src/main/scala/rise/core/equality.scala

```diff
@@ -155,7 +155,7 @@ object equality {
       typeEq.equiv[TypeKind](typeEnv)(a.t)(b.t) && (a match {
         case Identifier(na) => and { case Identifier(nb) => exprEnv.check(na, nb)}
         case Literal(da) => and { case Literal(db) => equivData(typeEnv)(da)(db) }
-        case a: Primitive => and { case b: Primitive => a.name == b.name }
+        case a: Primitive => and { case b: Primitive => a.getClass == b.getClass }
```

Comment:
should it not be `a == b` now then?

Reply:
otherwise if `a == b` does not work (if it also looks at the type) then instead of overriding `equals` we could have a separate `primitiveA.samePrimitive(primitiveB)` method?

Reply:
Actually, since some primitives might contain further expressions, should we not have this `samePrimitive` method take the equality environment generated by `=~=` so that it can keep checking further?

Reply:
Primitives should not contain further expressions, they are only parameterised by compile-time constants that are outside of the language itself.

Reply:
What about assertions, annotations and opaque expressions? Do we just assume that alpha equivalence will only run on expressions which had type inference ran on them and that all opaque expressions are equal?

Reply:
Do we actually still use type annotation primitives? This is what happens in the DSL:
```scala
  implicit class TypeAnnotation(t: Type) {
    def ::(e: Expr): Expr = e.setType(t)
    def `:`(e: Expr): Expr = e.setType(t)
  }
```
The primitive was used in a previous DSL which has been removed, so maybe we can remove the primitive too?
We already discussed that we should remove type assertions.
And finally for `Opaque` it really looks like it should not be a primitive, unless we can rewrite it into something like `def opaque: {ŧ: Type} -> t -> t` or `def opaque: {t1: Type} -> {t2: Type} -> t1 -> t2` if that is not too clunky.

Reply:
The code for type annotations that is used with `ToBeTyped` is this here:
```
implicit class TypeAnnotationHelper(t: Type) {
    def ::[T <: Expr](e: ToBeTyped[T]): ToBeTyped[Expr] =
      e >>= (e => toBeTyped(TypeAnnotation(e, t)))
  }
```

Reply:
Ah sorry about that, looks like I was looking at the `FSmooth` DSL code.

Reply:
Then I think we can close https://github.com/rise-lang/shine/issues/153 and think about  https://github.com/rise-lang/shine/issues/152.

Reply:
I think for the time being and until we pull them out of the set of primitives, I think it is warranted to have especial cases for these "primitives" in alpha equivalence?

Reply:
Sounds reasonable short-term.